source/common/common/packed_struct.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <type_traits>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+
+/**
+ * Helper class to pack elements of the same type into a struct like
+ * object, with the expectation that some or most of the elements are not set.
+ *
+ * The overhead of using this class as opposed to declaring the elements
+ * individually in another class or struct is 8 + (n+1) unaligned bytes, where n is
+ * the maximum number of elements in the struct.
+ *
+ * The way to use this class is to first define an enum class with the element
+ * names, and providing the enum class as the third template argument to
+ * PackedStruct, the first two being the type of element and number of element
+ * names. Instantiate the object with the number of elements that have values.
+ * Note that this can be increased after the fact but can (potentially) lead to
+ * memory fragmentation.
+ *
+ * For example:
+ *
+ * ```
+ *    enum class MyElementNames {
+ *      element0,
+ *      element1,
+ *      element2,
+ *      element3
+ *    };
+ *
+ *    using MyPackedStringStruct = PackedStruct<std::string, 4, MyElementNames>;
+ *
+ *    MyPackedStringStruct packed_struct(2); // For this instantiation, only 2
+ *                                           // elements are non-empty
+ *    packed_struct.set<MyElementNames::element1>("abc");
+ *    packed_struct.set<MyElementNames::element2>("def");
+ *
+ *    // Code that accesses the elements of the struct needs to check for existence first.
+ *    if (packed_struct.has<MyElementNames::element0>() {
+ *      auto& element0 = packed_struct.get<MyElements::element0>();
+ *    }
+ * ```
+ */
+template <class T, uint8_t max_size, class ElementName> class PackedStruct {
+public:
+  PackedStruct(size_t a_capacity = 0) : data_(a_capacity > 0 ? new T[a_capacity] : nullptr) {
+    static_assert(std::is_enum_v<ElementName>);
+    static_assert(max_size > 0);
+
+    RELEASE_ASSERT(a_capacity < std::numeric_limits<uint8_t>::max(),
+                   "capacity should fit in uint8_t.");
+    indices_.fill(max_size);
+    indices_[max_size] = a_capacity;
+  }
+
+  // Accessors.
+  template <ElementName element_name> T& get() {
+    sizeCheck<element_name>();
+    hasCheck<element_name>();
+    return (data_.get())[indices_[static_cast<std::underlying_type_t<ElementName>>(element_name)]];
+  }
+  template <ElementName element_name> const T& get() const {
```

Comment:
to remove duplicated code you could define the const get in terms of the non-const one with a const_cast. up to you.